openapi/swaggerui.go

```diff
@@ -0,0 +1,115 @@
+)
+
+const (
+	defaultSwaggerHost = "https://petstore3.swagger.io"
+	swaggerUITemplate  = `
+	<!-- HTML for static distribution bundle build -->
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8">
+    <title>API documentation</title>
+    <link rel="stylesheet" type="text/css" href="{{ .SwaggerHost }}/swagger-ui.css" >
+    <link rel="icon" type="image/png" href="{{ .SwaggerHost }}/favicon-32x32.png" sizes="32x32" />
+    <link rel="icon" type="image/png" href="{{ .SwaggerHost }}/favicon-16x16.png" sizes="16x16" />
+    <style>
+      html
+      {
+        box-sizing: border-box;
+        overflow: -moz-scrollbars-vertical;
+        overflow-y: scroll;
+      }
+      *,
+      *:before,
+      *:after
+      {
+        box-sizing: inherit;
+      }
+      body
+      {
+        margin:0;
+        background: #fafafa;
+      }
+    </style>
+  </head>
+  <body>
+    <div id="swagger-ui"></div>
+    <script src="{{ .SwaggerHost }}/swagger-ui-bundle.js"> </script>
+    <script src="{{ .SwaggerHost }}/swagger-ui-standalone-preset.js"> </script>
+    <script>
+    window.onload = function() {
+      // Begin Swagger UI call region
+      const ui = SwaggerUIBundle({
+        "dom_id": "#swagger-ui",
+        deepLinking: true,
+        presets: [
+          SwaggerUIBundle.presets.apis,
+          SwaggerUIStandalonePreset
+        ],
+        plugins: [
+          SwaggerUIBundle.plugins.DownloadUrl
+        ],
+        layout: "StandaloneLayout",
+        url: "{{ .SpecJSONPath }}",
+      })
+      // End Swagger UI call region
+      window.ui = ui
+    }
+  </script>
+  </body>
+</html>`
+)
+
+// SwaggerConfig configures the SwaggerDoc middlewares.
+type SwaggerConfig struct {
+	// DocPrefix the url to find the doc
+	DocPath string
+	// SpecJsonPath the url to find the spec
+	SpecJSONPath string
+	// SwaggerHost for the js that generates the swagger ui site, defaults to: http://petstore3.swagger.io/
+	SwaggerHost string
+}
+
+// NewSwaggerConfig return swaggerConfig.
+func NewSwaggerConfig(docPath, specJSONPath, swaggerHost string) *SwaggerConfig {
+	if swaggerHost == "" {
+		swaggerHost = defaultSwaggerHost
+	}
+	return &SwaggerConfig{
+		DocPath:      docPath,
+		SpecJSONPath: specJSONPath,
+		SwaggerHost:  swaggerHost,
+	}
+}
+
+// NewSwaggerDocUI creates a echo middleware to serve a documentation site for a swagger spec.
+// This allows for altering the spec before starting the http listener.
+func NewSwaggerDocUI(config *SwaggerConfig, swaggerJSON []byte) (echo.MiddlewareFunc, error) {
+	// swagger html
+	tmpl := template.Must(template.New("swaggerdoc").Parse(swaggerUITemplate))
+	buf := bytes.NewBuffer(nil)
+	err := tmpl.Execute(buf, config)
+	if err != nil {
+		return nil, err
+	}
+	uiHTML := buf.Bytes()
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			path := c.Request().URL.Path
+			switch {
+			case path == config.DocPath:
+				return c.HTML(200, string(uiHTML))
```

Comment:
if we returned here, how can chained `next` be applied?

Reply:
this is by desigin😬  when the case `path == config.DocPath` is hit, I want to then return the rendered document page

In cases other than this case, the middleware will call `next(c)` to ensure that the other middleware is called in a chain in L112

Reply:
if a "counting for access" handler is added to chain, we will lose counting if don't apply them.

anyway, I think generally we should always propagate the `echo.Context` in the chain. "return the rendered document page(raw html)" doesn't interfere with `next(c)`

Reply:
> if a "counting for access" handler is added to chain, we will lose counting if don't apply them.

so user need to add this counter mw befroe doc mw, in other words, this mw is already at the end of the request lifecycle and all it needs to do is write the http reponse back to the client.

If we don't put it in the last middleware of the request, it could happen that the document-related html is written to the response, but this will be overwritten by the subsequent middleware, which I don't think is the state the user wants

Reply:
In some cases HTML should be overwritten/appended, like append a JS script. But I don't think there's a case that chaining should be broken other than we meet a critical error.

Reply:
I think a direct return makes sense here, as the scenario is more like an interceptor than other chain mw

To give an example of the official echo jwt [mw](https://echo.labstack.com/middleware/jwt/) 
In the logic of this jwt mw, when an illegal jwt token is encountered, errror=invalid token is returned directly

Reply:
I don't think there's a case that chaining should be broken **other than we meet a critical error**.

Reply:
or i can use new http handelr to serve doc?https://github.com/pingcap/dm/commit/62298cb106d211b9b459156d6a96b53e6bdd5600
but this code feels ugly to me